### PR TITLE
Fix #239

### LINKED ISSUE #239: Applying metadata filtering when making vmrox=vmro3+vmrno2 does not work as expected
I need to be able to use the normal metadata filtering functionality when combining O3 and NO2 to OX. The basic functionality works as far as I can tell, but I seem to only be able to use the very small number of metadata keys transferred to StationData objects to filter by. Is this correct? I get an error message listing a much smaller number of valid keys than is normally available in GHOST.

I set it up like this:

```Python
'EEA-ox-rural'        :   dict(obs_id='TEST',
                              obs_vars=['vmrox'],
                              obs_type='ungridded',
                              obs_vert_type='Surface',
                              web_interface_name='EEA-rural',
                              ts_type='daily',
                              obs_merge_how={
                                  'vmrox' : 'eval',},
                              obs_aux_requires = {
                                  'vmrox' : {
                                      'GHOST.EEA.daily' : ['vmro3','vmrno2'],
                                      }},
                              obs_aux_funs = {
                                  'vmrox' :
                                      # variables used in computation method need to be based on AeroCom
                                      # units, since the colocated StationData objects (from which the
                                      # new UngriddedData is computed, will perform AeroCom unit check
                                      # and conversion)
                                      '(GHOST.EEA.daily;vmro3+GHOST.EEA.daily;vmrno2)'
                                      },
                                  obs_aux_units = {
                                      'vmrox' : 'nmol mol-1'
                                      },
                              ignore_station_names=ignore_dict,
                              obs_filters= {**standard_filter,**rural_filter,**mountain_filter}
                              ),
```

Where the obs_filter is set up like this and works when not using the new functionality:

```Python
no2_ignore_station_names = []
o3_ignore_station_names = []
pm10_ignore_station_names = []
pm25_ignore_station_names = []

ignore_dict = {'concno2':no2_ignore_station_names,
               'vmro3':o3_ignore_station_names,
               'concpm10':pm10_ignore_station_names,
               'concpm25':pm25_ignore_station_names}

MBlandforms_to_include = ['high altitude plains','water', 'very low plateaus',
                        'plains', 'rugged lowlands','hills','high altitude plateaus',
                        'mid altitude plateaus', 'nan','lowlands', 'mid altitude plains',
                        'low plateaus',]

EEA_rural_station_types_to_include = ['background']
EEA_rural_area_types_to_include = ['rural','rural-near_city','rural-regional', 'rural-remote']


#Define filters for the obs subsets

standard_filter = {'set_flags_nan':True,}

rural_filter = {'standardised_network_provided_station_classification':EEA_rural_station_types_to_include,
                'standardised_network_provided_area_classification':EEA_rural_area_types_to_include}

mountain_filter = {'altitude':[-20,1500],
                   'ESDAC_Meybeck_landform_classification':MBlandforms_to_include}
```

### PR DESCRIPTION
**Pull request by jgliss**
Attempt to increase functionality to enable flexible metadata filtering also for auxiliary UngriddedData objects. Will come with some restructuring of where metadata filters of UngriddedData objects are applied in the highlevel colocation routines.

### CODE REVIEW

**Comment by hansbrenna** [comment]
Do I understand correctly that the flow is like this: 

When you specify filters to filter_by_meta, these meta keys will be transferred to the collection of StationData objects produced by the ungridded-ungridded colocation. Then the station data objects are filtered by meta afterwards.

**Comment by jgliss** [comment]
> Do I understand correctly that the flow is like this:
> 
> When you specify filters to filter_by_meta, these meta keys will be transferred to the collection of StationData objects produced by the ungridded-ungridded colocation. Then the station data objects are filtered by meta afterwards.

@hansbrenna no, filter_by_meta operates directly on the UngriddedData object, so first, vmro3 and and vmrno2 are read, then each of them is filtered based on your site classification, and then these 2 UngriddedData objects are colocated in space and time and based on these, vmrox is computed for common sites and put into a new UngriddedData object, which is then, in the colocation routine, converted to list of StationData objects, which is added to the ColocatedData objects.

**Comment by jgliss** [comment]
@hansbrenna I think this is ready for review now. I found and fixed some bugs today related to `ignore_station_names` filtering, which may also affect your analysis (I remember you merged this branch a few days ago into one of yours to test the vmrox stuff). Can you review, please (and update your current working branch)? Please let me know if you cannot review, so I can ask @ejgal. Cheers

**Comment by jgliss** [comment]
@hansbrenna I fixed a bug in applying metadata filters when filter is a list (like your rural filters). Can you merge and check if the filters are applied properly now?

**Comment by hansbrenna** [comment]
I'll test now.

**Comment by jgliss** [comment]
@hansbrenna I guess you are too busy to review this, I really would like to merge it though, because I have too many branches in parallel at the moment. @ejgal could you have a look at it, please?

File: pyaerocom/combine_vardata_ungridded.py [hunk 1]
```diff
@@ -290,7 +292,7 @@ def combine_vardata_ungridded(data_ids_and_vars,
     Note
     ----
     Be aware of unit conversion issues that may arise if your input data is
-    not in AeroCom default units. For details see below.
+    not in AeroCom default units. For that reason,  For details see below.
```
Comment (ejgal) [inline pyaerocom/combine_vardata_ungridded.py:295]:
```suggestion
    not in AeroCom default units. For details see below.
```

File: pyaerocom/ungriddeddata.py [hunk 2]
```diff
@@ -1169,28 +1179,98 @@ def get_variable_data(self, variables, start=None, stop=None,
...
-                    if not fnmatch.fnmatch(meta[k], v):
+        if not self._check_str_filter_match(meta, negate, str_f):
+            return False
+
+        for metakey, filterval in list_f.items():
+            if not metakey in meta:
+                return False
+            if metakey == 'station_name':
+                statname = meta['station_name']
+                if any([statname.startswith(x) for x in ['Tr', 'Mauna']]):
```
Comment (ejgal) [inline pyaerocom/ungriddeddata.py:1230]: Debugging?

```suggestion
```

File: pyaerocom/ungriddeddata.py [hunk 3]
```diff
@@ -1169,28 +1179,98 @@ def get_variable_data(self, variables, start=None, stop=None,
...
+                    print(statname)
```
Comment (ejgal) [inline pyaerocom/ungriddeddata.py:1231]: Debugging?
```suggestion
```

File: pyaerocom/ungriddeddata.py [hunk 4]
```diff
@@ -1169,28 +1179,98 @@ def get_variable_data(self, variables, start=None, stop=None,
...
-                if '*' in v:
-                    if not fnmatch.fnmatch(meta[k], v):
+        if not self._check_str_filter_match(meta, negate, str_f):
+            return False
+
+        for metakey, filterval in list_f.items():
+            if not metakey in meta:
+                return False
+            if metakey == 'station_name':
+                statname = meta['station_name']
```
Comment (ejgal) [inline pyaerocom/ungriddeddata.py:1229]: This is not used anywhere else, right?
```suggestion
```

**Review by ejgal — APPROVED** [review]
Good to go, but please check some minor things in suggestions.
Also, you can trigger a rerun of the tests without adding an empty commit:
Checks -> CI, on:push -> Rerun jobs

**Comment by jgliss** [comment]
> Good to go, but please check some minor things in suggestions.
> Also, you can trigger a rerun of the tests without adding an empty commit:
> Checks -> CI, on:push -> Rerun jobs

Thanks!!
